app/controllers/test/oidc_test_controller.rb

```diff
@@ -0,0 +1,154 @@
+      update_service_provider if @update_redirect_uri
+    end
+
+    def auth_request
+      ial = prepare_step_up_flow(ial: params[:ial])
+
+      idp_url = authorization_url(
+        ial: ial,
+        aal: params[:aal],
+      )
+
+      Rails.logger.info("Redirecting to #{idp_url}")
+
+      redirect_to(idp_url)
+    end
+
+    def auth_result
+      redirect_to('/')
+    end
+
+    def logout
+      redirect_to(logout_uri)
+    end
+
+    def authorization_url(ial:, aal: nil)
+      authorization_endpoint = openid_configuration[:authorization_endpoint]
+      params = ial2_params(
+        client_id: client_id,
+        acr_values: acr_values(ial: ial, aal: aal),
+        biometric_comparison_required: ial == BIOMETRIC_REQUIRED,
+        state: random_value,
+        nonce: random_value,
+      )
+      request_params = params.merge(
+        {
+          scope: scopes_for(ial),
+          redirect_uri: test_oidc_auth_result_url,
+        },
+      ).compact.to_query
+      "#{authorization_endpoint}?#{request_params}"
+    end
+
+    def prepare_step_up_flow(ial:)
+      if ial == 'step-up'
+        ial = '1'
+      end
+      ial
+    end
+
+    def scopes_for(ial)
+      case ial
+      when '0'
+        'openid email social_security_number'
+      when '1', nil
+        'openid email'
+      when '2', BIOMETRIC_REQUIRED
+        'openid email profile social_security_number phone address'
+      else
+        raise ArgumentError.new("Unexpected IAL: #{ial.inspect}")
+      end
+    end
+
+    def acr_values(ial:, aal:)
+      values = []
+
+      values << {
+        '0' => 'http://idmanagement.gov/ns/assurance/ial/0',
+        nil => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '' => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '1' => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '2' => 'http://idmanagement.gov/ns/assurance/ial/2',
+        'biometric-comparison-required' => 'http://idmanagement.gov/ns/assurance/ial/2',
+      }[ial]
+
+      values << {
+        '2' => 'http://idmanagement.gov/ns/assurance/aal/2',
+        '2-phishing_resistant' => 'http://idmanagement.gov/ns/assurance/aal/2?phishing_resistant=true',
+        '2-hspd12' => 'http://idmanagement.gov/ns/assurance/aal/2?hspd12=true',
+      }[aal]
+
+      values.compact.join(' ')
+    end
+
+    def json(response)
+      JSON.parse(response.to_s).with_indifferent_access
+    end
+
+    def random_value
+      SecureRandom.hex
+    end
+
+    def client_id
+      @client_id
+    end
+
+    private
+
+    def logout_uri
+      endpoint = openid_configuration[:end_session_endpoint]
+      request_params = {
+        client_id: client_id,
+        post_logout_redirect_uri: '/',
+        state: SecureRandom.hex,
+      }.to_query
+
+      "#{endpoint}?#{request_params}"
+    end
+
+    def openid_configuration
+      @openid_configuration ||= OpenidConnectConfigurationPresenter.new.configuration
```

Comment:
This is can be fetched at {root_url}/.well-known/openid-configuration, but due to Puma dev mode and RAILS_MAX_THREAD configuration(with value of 1), making a loop back request to the same server will cause deadlock.